src/coprocessor/dag/dag.rs

```diff
@@ -92,27 +89,114 @@ impl DAGContext {
                     let mut resp = Response::new();
                     let mut sel_resp = SelectResponse::new();
                     sel_resp.set_chunks(RepeatedField::from_vec(chunks));
-                    let mut counts = Vec::with_capacity(4);
-                    self.exec.collect_output_counts(&mut counts);
-                    sel_resp.set_output_counts(counts);
+                    self.exec
+                        .collect_output_counts(sel_resp.mut_output_counts());
                     let data = box_try!(sel_resp.write_to_bytes());
                     resp.set_data(data);
                     return Ok(resp);
                 }
-                Err(e) => if let Error::Other(_) = e {
-                    let mut resp = Response::new();
-                    let mut sel_resp = SelectResponse::new();
-                    sel_resp.set_error(to_pb_error(&e));
-                    resp.set_data(box_try!(sel_resp.write_to_bytes()));
-                    resp.set_other_error(format!("{}", e));
-                    return Ok(resp);
-                } else {
-                    return Err(e);
+                Err(e) => match e {
+                    Error::Other(_) => {
+                        let mut resp = Response::new();
+                        let mut sel_resp = SelectResponse::new();
+                        sel_resp.set_error(to_pb_error(&e));
+                        resp.set_data(box_try!(sel_resp.write_to_bytes()));
+                        resp.set_other_error(format!("{}", e));
+                        return Ok(resp);
+                    }
+                    _ => return Err(e),
                 },
             }
         }
     }
 
+    pub fn handle_streaming_request(
+        &mut self,
+        batch_row_limit: usize,
+    ) -> Result<(Option<Response>, bool)> {
+        let (mut record_cnt, mut finished) = (0, false);
+        let mut chunk = Chunk::new();
+        let (mut start_key, mut support_partial_retry) = (None, true);
+        while record_cnt < batch_row_limit {
+            match self.exec.next() {
+                Ok(Some(row)) => {
+                    record_cnt += 1;
+                    if support_partial_retry && start_key.is_none() {
+                        start_key = self.exec.take_last_key();
+                        if start_key.is_none() {
+                            support_partial_retry = false;
+                        }
+                    }
```

Comment:
It seems you only set start_key in the first time which means `record_cnt==1`, so maybe use the following instead of L124-129 could be much easier to understand?
```
    if record_cnt == 1 {
      start_key = self.exec.take_last_key();
   }
````